src/internal_events/statsd.rs

```diff
@@ -93,3 +96,29 @@ impl<T: std::fmt::Debug + std::fmt::Display> InternalEvent for StatsdSocketError
         );
     }
 }
+
+#[derive(Debug)]
+pub struct StatsdInvalidMetric<'a> {
```

Comment:
I note that [our RFC](https://github.com/timberio/vector/blob/eaf56397175a54e6a18cbc62966d43ecb926b917/rfcs/2020-03-17-2064-event-driven-observability.md) describes:

> Both event struct names and metric keys follow the `{noun}_{verb}` naming scheme. This keeps things simple, and we can rely on tags/labels for more context.

(for structs, this would follow Rust's conventions as `{Noun}{Verb}`)

I see that we aren't applying that rule universally, as we also have an `StatsdInvalidRecord` event.

I guess we could name this `StatsdInvalidMetricSent`? Although from what I understand, this event actually describes the sink _receiving_ an invalid event, and then not _sending_ it to Statsd, so that name also doesn't describe what happened.

You could name it `StatsdInvalidMetricReceived`, but then you'd potentially be getting in the way of the statsd source.

This makes me wonder if we should either:

1. distinguish between sources and sinks of the same type by using `Statsd{Sink,Source}{...}` event names.
2. use the same event, but have `component_kind` be dynamically set to either `sink` or `source`

I'd vote for the first solution (unless there is another, better one), as the second one can get quite messy if you need to add different labels for the source and sink of the same type.

cc @lukesteensen

Reply:
I think the first solution makes more sense to me, sources and sinks are completely different concepts, and there is often very little shared between them, as they both handle different sides of the same protocol.

I was also thinking it may be worth separating the events into separate modules, so a `internal_events/statsd_source.rs` and a `internal_events/statsd_sink.rs` or similar. It is certainly possible to compile Vector with just the statsd sink and not the source.  The extra feature flags I have had to put into the module suggests that lumping it all in one module is a bit of an anti pattern.

Reply:
Yeah, I agree with your insight regarding the feature flags. Having a separate module for each component does feel like the right solution 👍

It is still somewhat orthogonal to the name of the events though, depending on how we want to use these internal events:

```rust
use internal_event::statsd_source::StatsdSourceInvalidMetricReceived;
use internal_event::statsd_sink::StatsdSinkInvalidMetricReceived;

emit!(StatsdSourceInvalidMetricReceived)
emit!(StatsdSinkInvalidMetricReceived)
```

vs.

```rust
use internal_event::{statsd_source, statsd_sink};

emit!(statsd_source::InvalidMetricReceived)
emit!(statsd_sink::InvalidMetricReceived)
```

I'd vote for the latter, although I can also see how that could potentially interfere with any actual non-internal-event `statd_source` module. A solution to that would be to always re-export internal events from the actual component modules. But I think as long as we can get away with it, without creating name conflicts, I'd vote to keep it simple and use the second example from above.

Reply:
At the risk of complicating things further by throwing too many options in the mix, what about something like:

```rust
emit!(InvalidMetricReceived(Source("statsd")))
emit!(InvalidMetricReceived(Sink("statsd")))
```

That way there is only one InvalidMetricRecieved event, and it is tagged with further information to indicate the source of the event. That way there is no code duplication, so we can ensure things stay consistent.

Admittedly, it would be more work to change all the events in line with that approach.

Reply:
I'm not terribly worried about the name of the event itself since that's not user-facing, but I agree it'd be nice to be unambiguous. Since only sources receive things, `StatsdInvalidMetricReceived` seems fine to me.

Reply:
Wait, now I'm confused;

> Since only sources receive things, `StatsdInvalidMetricReceived` seems fine to me.

This would imply to me that you'd be fine with that name, _if_ we used it for the statsd source, right? Because right now, we're using this event in the sink, which still seems confusing to me.